### ISSUE TITLE
Data sources errors are not self-descriptive in the nature of the error

### ISSUE THREAD

**Issue by HebertCL** [hashicorp/terraform-provider-tfe#1309]
## Context
I have a Terraform configuration that have worked for the past year or so (I didn't create it but rather inherit it as part of me joining the team), in which part of the first things I did was deprecate an old API key and setup a new one.
The whole workspace creation and maintenance is done in a workspace which all of the sudden stopped working. This workspace configuration resides in a separated repository and it handles the rest of the workspace's lifecycle.

#### Use-cases

Recently I've been retiring old, unused workspaces along with the creation of new ones. My configuration uses at least `tfe_organization` and `tfe_workspace_ids` data sources which I use in combination of other resources to automate the creation of workspace, assignment of variable sets and creation of variables.

Since at least 7 days ago I encountered all my plans, either from CLI or TFC console returned the following error:

```sh
data.tfe_organization.this: Reading...
data.tfe_workspace_ids.workspace_ids["TFC-CREDS"]: Reading...
data.tfe_workspace_ids.workspace_ids["GCP-CREDS"]: Reading...

Planning failed. Terraform encountered an error while generating this plan.

╷
│ Error: could not read organization 'my-org'
│ 
│   with data.tfe_organization.this,
│   on main.tf line 9, in data "tfe_organization" "this":
│    9: data "tfe_organization" "this" {
│ 
╵
╷
│ Error: Error retrieving workspaces: resource not found
│ 
│   with data.tfe_workspace_ids.workspace_ids["GCP-CREDS"],
│   on main.tf line 13, in data "tfe_workspace_ids" "workspace_ids":
│   13: data "tfe_workspace_ids" "workspace_ids" {
│ 
╵
╷
│ Error: Error retrieving workspaces: resource not found
│ 
│   with data.tfe_workspace_ids.workspace_ids["TFC-CREDS"],
│   on main.tf line 13, in data "tfe_workspace_ids" "workspace_ids":
│   13: data "tfe_workspace_ids" "workspace_ids" {
```

The token I set doesn't have expiration date. I tried using `TF_LOG=DEBUG` when running plans in the CLI, but couldn't get any additional information about this errors.

#### Attempted Solutions

I decided to run several tests to rule out every potential issue I could think of. That ended up including issues with the token I was using.
First, I tried using a token created using `terraform login` at CLI level. This token worked just fine and found no issues calling data sources.
Next up, I took that token and intentionally introduced typos to see what error I would get in the scenario of using an invalid token. This is what I got:

```sh
data.tfe_organization.this: Reading...
data.tfe_workspace_ids.workspace_ids["TFC-CREDS"]: Reading...
data.tfe_workspace_ids.workspace_ids["GCP-CREDS"]: Reading...

Planning failed. Terraform encountered an error while generating this plan.

╷
│ Error: Error retrieving organization: unauthorized
│ 
│   with data.tfe_organization.this,
│   on main.tf line 9, in data "tfe_organization" "this":
│    9: data "tfe_organization" "this" {
│ 
╵
╷
│ Error: Error retrieving organization: unauthorized
│ 
│   with data.tfe_workspace_ids.workspace_ids["GCP-CREDS"],
│   on main.tf line 13, in data "tfe_workspace_ids" "workspace_ids":
│   13: data "tfe_workspace_ids" "workspace_ids" {
│ 
╵
╷
│ Error: Error retrieving organization: unauthorized
│ 
│   with data.tfe_workspace_ids.workspace_ids["TFC-CREDS"],
│   on main.tf line 13, in data "tfe_workspace_ids" "workspace_ids":
│   13: data "tfe_workspace_ids" "workspace_ids" {
```
Next up, I tried with a token only valid for 1 hour. I made this assumption thinking that the current token giving me grief was indeed expired, and that another expired token would switch from working normally to the `resource not found` error.
Token indeed worked fine for 1 hour, and by using it up next I encountered with the `unauthorized` error that the invalid token was giving me.

In conclusion, I wasn't able to figure out why my supposedly valid token would give me a `resource not found` error, but the only clear workaround was replacing it for a brand new token.

#### Proposal

Introducing self-descriptive errors, that allow you to get a more detailed explanation whenever you enable the debug variable would be highly appreciated.

To me `unauthorized` give me a hint of what's going wrong with my token. However, `resource not found` is not something I can use to figure out what is not working and how to solve it, specially given the fact that, in my case, it was a token issue where such token belongs to the same org, was created by the same user with the same level of permissions.

**Comment by brandonc** [comment]
@HebertCL Hey, thanks for all the detail that you included when reporting this issue.

There is a crucial nuance between these error messages that is not at all obvious! Please see this documentation for details https://developer.hashicorp.com/terraform/cloud-docs/api-docs#authentication

In short, unauthorized is returned when a token is problematic, like being malformed or expired. resource not found is used both for actual not found errors but also in place of 403 (forbidden) errors. This serves to not leak information about the existence of resources that a user is not authorized to read. If we were to change the API to return details about the token not being able to access the requested resource, but a different error when the resource did not actually exist, an attacker could infer information about what they don't have access to.

Hope this helps! I've learned over time to initially interpret 404 errors in terraform as authorization errors because actual resource not found errors are rare and often ignored (for example, when deleting a resource that has already been deleted)
